File: commands/wrapped.js
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../database/db.js');
const logger = require('../logger');

function formatTime(totalSeconds) {
    const hours = Math.floor(totalSeconds / 3600);
    const minutes = Math.floor((totalSeconds % 3600) / 60);

    if (hours >= 24) {
        const days = Math.floor(hours / 24);
        const remainingHours = hours % 24;
        return `${days}d ${remainingHours}h ${minutes}m`;
    }
    return `${hours}h ${minutes}m`;
}

function formatHour(hour) {
    if (hour === 0) return '12 AM';
    if (hour === 12) return '12 PM';
    return hour < 12 ? `${hour} AM` : `${hour - 12} PM`;
}

module.exports = {
    data: new SlashCommandBuilder()
        .setName('wrapped')
        .setDescription('View music listening stats')
        .addSubcommand(sub =>
            sub.setName('me')
                .setDescription('Your personal music stats')
        )
        .addSubcommand(sub =>
            sub.setName('server')
                .setDescription('Server-wide music stats')
        )
        .addSubcommand(sub =>
            sub.setName('compare')
                .setDescription('Compare music taste with another user')
                .addUserOption(opt =>
                    opt.setName('user')
                        .setDescription('User to compare with')
                        .setRequired(true)
                )
        ),

    async execute(interaction) {
        const subcommand = interaction.options.getSubcommand();
        const emotes = interaction.client.emotes;

        try {
            if (subcommand === 'me') {
                await handleMe(interaction, emotes);
            } else if (subcommand === 'server') {
                await handleServer(interaction, emotes);
            } else if (subcommand === 'compare') {
                await handleCompare(interaction, emotes);
            }
        } catch (e) {
            logger.error('Wrapped command error', { error: e.message, stack: e.stack });
            const embed = new EmbedBuilder()
                .setDescription(`${emotes.error} | Something went wrong fetching stats.`)
                .setColor('#ff0000')
                .setFooter({ text: 'The Pack', iconURL: interaction.client.logo });
            return interaction.editReply({ embeds: [embed] });
        }
    },
};

async function handleMe(interaction, emotes) {
    const guildId = interaction.guildId;
    const userId = interaction.user.id;

    const [[stats]] = await db.pool.query(
        `SELECT COUNT(*) as totalTracks,
                COALESCE(SUM(durationSeconds), 0) as totalSeconds,
                COUNT(DISTINCT CONCAT(trackTitle, trackArtist)) as uniqueTracks
         FROM ListeningHistory WHERE guildId = ? AND odUserId = ?`,
        [guildId, userId]
    );

    if (parseInt(stats.totalTracks) === 0) {
        const embed = new EmbedBuilder()
            .setDescription(`${emotes.error} | No listening data found for you in this server yet!`)
            .setColor('#ff0000')
            .setFooter({ text: 'The Pack', iconURL: interaction.client.logo });
        return interaction.editReply({ embeds: [embed] });
    }

    const [topTracks] = await db.pool.query(
        `SELECT trackTitle as title, trackArtist as artist, COUNT(*) as playCount
         FROM ListeningHistory WHERE guildId = ? AND odUserId = ?
         GROUP BY trackTitle, trackArtist ORDER BY playCount DESC LIMIT 5`,
        [guildId, userId]
    );

    const [topArtists] = await db.pool.query(
        `SELECT trackArtist as artist, COUNT(*) as playCount
         FROM ListeningHistory WHERE guildId = ? AND odUserId = ? AND trackArtist IS NOT NULL
         GROUP BY trackArtist ORDER BY playCount DESC LIMIT 5`,
        [guildId, userId]
    );

    const [favoriteHour] = await db.pool.query(
        `SELECT HOUR(playedAt) as hour, COUNT(*) as count
         FROM ListeningHistory WHERE guildId = ? AND odUserId = ?
         GROUP BY HOUR(playedAt) ORDER BY count DESC LIMIT 1`,
        [guildId, userId]
    );

    const tracksText = topTracks.map((t, i) => {
        const medal = i === 0 ? '🥇' : i === 1 ? '🥈' : i === 2 ? '🥉' : `**${i + 1}.**`;
        return `${medal} **${t.title}** — ${t.artist || 'Unknown'} (${t.playCount} plays)`;
    }).join('\n');

    const artistsText = topArtists.map((a, i) => {
        const medal = i === 0 ? '🥇' : i === 1 ? '🥈' : i === 2 ? '🥉' : `**${i + 1}.**`;
        return `${medal} **${a.artist}** (${a.playCount} plays)`;
    }).join('\n');

    const embed = new EmbedBuilder()
        .setTitle(`🎧 ${interaction.user.username}'s Music Wrapped`)
        .setThumbnail(interaction.user.displayAvatarURL())
        .addFields(
            { name: 'Total Listening Time', value: formatTime(stats.totalSeconds), inline: true },
            { name: 'Tracks Played', value: `${stats.totalTracks}`, inline: true },
            { name: 'Unique Tracks', value: `${stats.uniqueTracks}`, inline: true },
            { name: '🎵 Top Tracks', value: tracksText || 'No data', inline: false },
            { name: '🎤 Top Artists', value: artistsText || 'No data', inline: false }
        )
        .setColor('#ff006a')
        .setFooter({ text: 'The Pack', iconURL: interaction.client.logo })
        .setTimestamp();

    if (favoriteHour.length > 0) {
        embed.addFields({
            name: '⏰ Peak Listening Hour',
            value: formatHour(favoriteHour[0].hour),
            inline: true
        });
    }

    return interaction.editReply({ embeds: [embed] });
}

async function handleServer(interaction, emotes) {
    const guildId = interaction.guildId;

    const [[stats]] = await db.pool.query(
        `SELECT COUNT(*) as totalTracks,
                COALESCE(SUM(durationSeconds), 0) as totalSeconds,
                COUNT(DISTINCT trackArtist) as uniqueArtists
         FROM ListeningHistory WHERE guildId = ?`,
        [guildId]
    );

    if (parseInt(stats.totalTracks) === 0) {
        const embed = new EmbedBuilder()
            .setDescription(`${emotes.error} | No listening data found for this server yet!`)
            .setColor('#ff0000')
            .setFooter({ text: 'The Pack', iconURL: interaction.client.logo });
        return interaction.editReply({ embeds: [embed] });
    }

    const [topTracks] = await db.pool.query(
        `SELECT trackTitle as title, trackArtist as artist, COUNT(*) as playCount
         FROM ListeningHistory WHERE guildId = ?
         GROUP BY trackTitle, trackArtist ORDER BY playCount DESC LIMIT 5`,
        [guildId]
    );

    const [topListeners] = await db.pool.query(
        `SELECT odUserId, COALESCE(SUM(durationSeconds), 0) as totalSeconds, COUNT(*) as playCount
         FROM ListeningHistory WHERE guildId = ?
         GROUP BY odUserId ORDER BY totalSeconds DESC LIMIT 5`,
        [guildId]
    );

    const tracksText = topTracks.map((t, i) => {
        const medal = i === 0 ? '🥇' : i === 1 ? '🥈' : i === 2 ? '🥉' : `**${i + 1}.**`;
        return `${medal} **${t.title}** — ${t.artist || 'Unknown'} (${t.playCount} plays)`;
    }).join('\n');

    const listenersText = topListeners.map((l, i) => {
        const medal = i === 0 ? '🥇' : i === 1 ? '🥈' : i === 2 ? '🥉' : `**${i + 1}.**`;
        return `${medal} <@${l.odUserId}> — ${formatTime(l.totalSeconds)} (${l.playCount} plays)`;
    }).join('\n');

    const embed = new EmbedBuilder()
        .setTitle(`🎧 ${interaction.guild.name} Music Wrapped`)
        .setThumbnail(interaction.guild.iconURL())
        .addFields(
            { name: 'Total Listening Time', value: formatTime(stats.totalSeconds), inline: true },
            { name: 'Tracks Played', value: `${stats.totalTracks}`, inline: true },
            { name: 'Unique Artists', value: `${stats.uniqueArtists}`, inline: true },
            { name: '🎵 Top Tracks', value: tracksText || 'No data', inline: false },
            { name: '👑 Top Listeners', value: listenersText || 'No data', inline: false }
        )
        .setColor('#ff006a')
        .setFooter({ text: 'The Pack', iconURL: interaction.client.logo })
        .setTimestamp();

    return interaction.editReply({ embeds: [embed] });
}

async function handleCompare(interaction, emotes) {
    const guildId = interaction.guildId;
    const otherUser = interaction.options.getUser('user');

    if (otherUser.id === interaction.user.id) {
        const embed = new EmbedBuilder()
            .setDescription(`${emotes.error} | You can't compare with yourself!`)
            .setColor('#ff0000')
            .setFooter({ text: 'The Pack', iconURL: interaction.client.logo });
        return interaction.editReply({ embeds: [embed] });
    }

    // Get top 50 tracks and 30 artists for each user
    const getUserData = async (uid) => {
        const [tracks] = await db.pool.query(
            `SELECT trackTitle as title, trackArtist as artist, COUNT(*) as playCount
             FROM ListeningHistory WHERE guildId = ? AND odUserId = ?
             GROUP BY trackTitle, trackArtist ORDER BY playCount DESC LIMIT 50`,
            [guildId, uid]
        );
        const [artists] = await db.pool.query(
            `SELECT trackArtist as artist, COUNT(*) as playCount
             FROM ListeningHistory WHERE guildId = ? AND odUserId = ? AND trackArtist IS NOT NULL
             GROUP BY trackArtist ORDER BY playCount DESC LIMIT 30`,
            [guildId, uid]
        );
        const [[stats]] = await db.pool.query(
            `SELECT COUNT(*) as totalTracks, COALESCE(SUM(durationSeconds), 0) as totalSeconds
             FROM ListeningHistory WHERE guildId = ? AND odUserId = ?`,
            [guildId, uid]
        );
        return { tracks, artists, stats };
    };

    const [u1, u2] = await Promise.all([
        getUserData(interaction.user.id),
        getUserData(otherUser.id)
    ]);

    if (parseInt(u1.stats.totalTracks) === 0 || parseInt(u2.stats.totalTracks) === 0) {
        const embed = new EmbedBuilder()
            .setDescription(`${emotes.error} | Both users need listening history to compare.`)
            .setColor('#ff0000')
            .setFooter({ text: 'The Pack', iconURL: interaction.client.logo });
        return interaction.editReply({ embeds: [embed] });
    }

    // Compute shared artists
    const artists1 = new Set(u1.artists.map(a => a.artist.toLowerCase()));
    const artists2 = new Set(u2.artists.map(a => a.artist.toLowerCase()));
    const sharedArtists = [...artists1].filter(a => artists2.has(a));

    // Compute shared tracks
    const tracks1 = new Set(u1.tracks.map(t => `${t.title}::${t.artist}`.toLowerCase()));
    const tracks2 = new Set(u2.tracks.map(t => `${t.title}::${t.artist}`.toLowerCase()));
    const sharedTracks = [...tracks1].filter(t => tracks2.has(t));

    const minArtists = Math.min(artists1.size, artists2.size);
    const compatibility = minArtists > 0 ? Math.round((sharedArtists.length / minArtists) * 100) : 0;

    // Pick emoji for compatibility level
    let compatEmoji = '💔';
    if (compatibility >= 80) compatEmoji = '❤️‍🔥';
    else if (compatibility >= 60) compatEmoji = '💖';
    else if (compatibility >= 40) compatEmoji = '💛';
    else if (compatibility >= 20) compatEmoji = '💙';

    const sharedArtistsText = sharedArtists.length > 0
        ? sharedArtists.slice(0, 10).join(', ')
        : 'None';

    const embed = new EmbedBuilder()
        .setTitle(`${compatEmoji} Music Compatibility: ${compatibility}%`)
        .setDescription(`**${interaction.user.username}** vs **${otherUser.username}**`)
        .addFields(
            { name: `${interaction.user.username}`, value: `${formatTime(u1.stats.totalSeconds)}\n${u1.stats.totalTracks} plays`, inline: true },
            { name: `${otherUser.username}`, value: `${formatTime(u2.stats.totalSeconds)}\n${u2.stats.totalTracks} plays`, inline: true },
            { name: '\u200B', value: '\u200B', inline: true },
            { name: '🎤 Shared Artists', value: sharedArtistsText, inline: false },
            { name: '📊 Stats', value: `**${sharedArtists.length}** shared artists • **${sharedTracks.length}** shared tracks`, inline: false }
        )
        .setColor('#ff006a')
        .setFooter({ text: 'The Pack', iconURL: interaction.client.logo })
        .setTimestamp();

    return interaction.editReply({ embeds: [embed] });
}

File: internal/commands/wrapped.go
package commands

import (
	"context"
	"fmt"
	"strings"

	"github.com/bwmarrin/discordgo"

	"github.com/OlliePCK/packbot/internal/storage"
	"github.com/OlliePCK/packbot/internal/style"
)

// compatibility computes the shared-artist percentage between two artist
// lists (case-insensitive), parity with the Node /wrapped compare math:
// shared ÷ min(len(a), len(b)) × 100.
func compatibility(artistsA, artistsB []string) (percent int, shared []string) {
	setA := make(map[string]bool, len(artistsA))
	for _, a := range artistsA {
		setA[strings.ToLower(a)] = true
	}
	setB := make(map[string]bool, len(artistsB))
	for _, b := range artistsB {
		setB[strings.ToLower(b)] = true
	}
	for a := range setA {
		if setB[a] {
			shared = append(shared, a)
		}
	}
	minSize := min(len(setA), len(setB))
	if minSize == 0 {
		return 0, shared
	}
	return int(float64(len(shared))/float64(minSize)*100 + 0.5), shared
}

func compatEmoji(percent int) string {
	switch {
	case percent >= 80:
		return "❤️‍🔥"
	case percent >= 60:
		return "💖"
	case percent >= 40:
		return "💛"
	case percent >= 20:
		return "💙"
	default:
		return "💔"
	}
}

// Wrapped is /wrapped — Spotify-Wrapped-style listening stats.
func Wrapped(d Deps) *Command {
	return &Command{
		Def: &discordgo.ApplicationCommand{
			Name:        "wrapped",
			Description: "View music listening stats",
			Options: []*discordgo.ApplicationCommandOption{
				{Type: discordgo.ApplicationCommandOptionSubCommand, Name: "me", Description: "Your personal music stats"},
				{Type: discordgo.ApplicationCommandOptionSubCommand, Name: "server", Description: "Server-wide music stats"},
				{
					Type: discordgo.ApplicationCommandOptionSubCommand, Name: "compare",
					Description: "Compare music taste with another user",
					Options: []*discordgo.ApplicationCommandOption{{
						Type: discordgo.ApplicationCommandOptionUser, Name: "user",
						Description: "User to compare with", Required: true,
					}},
				},
			},
		},
		Run: func(ctx context.Context, s *discordgo.Session, i *discordgo.InteractionCreate) error {
			sub, opts := subcommand(i)
			om := optionMap(opts)
			user := interactionUser(i)

			switch sub {
			case "me":
				return wrappedMe(ctx, d, s, i, user)
			case "server":
				return wrappedServer(ctx, d, s, i)
			case "compare":
				other := om["user"].UserValue(s)
				return wrappedCompare(ctx, d, s, i, user, other)
			}
			return fmt.Errorf("unknown wrapped subcommand %q", sub)
		},
	}
}

func trackLines(tracks []storage.TrackPlays) string {
	var lines []string
	for idx, t := range tracks {
		artist := t.Artist
		if artist == "" {
			artist = "Unknown"
		}
		lines = append(lines, fmt.Sprintf("%s **%s** — %s (%d plays)", medal(idx), t.Title, artist, t.PlayCount))
	}
	if len(lines) == 0 {
		return "No data"
	}
	return strings.Join(lines, "\n")
}

func artistLines(artists []storage.ArtistPlays) string {
	var lines []string
	for idx, a := range artists {
		lines = append(lines, fmt.Sprintf("%s **%s** (%d plays)", medal(idx), a.Artist, a.PlayCount))
	}
	if len(lines) == 0 {
		return "No data"
	}
	return strings.Join(lines, "\n")
}

func wrappedMe(ctx context.Context, d Deps, s *discordgo.Session, i *discordgo.InteractionCreate, user *discordgo.User) error {
	stats, err := d.Store.UserWrappedStats(ctx, i.GuildID, user.ID)
	if err != nil {
		return err
	}
	if stats.TotalTracks == 0 {
		return Respond(s, i, style.ErrorEmbed("No listening data found for you in this server yet!"))
	}

	topTracks, err := d.Store.TopTracks(ctx, i.GuildID, user.ID, 5)
	if err != nil {
		return err
	}
	topArtists, err := d.Store.TopArtists(ctx, i.GuildID, user.ID, 5)
	if err != nil {
		return err
	}
	hour, err := d.Store.FavoriteHour(ctx, i.GuildID, user.ID)
	if err != nil {
		return err
	}

	embed := &discordgo.MessageEmbed{
		Title:     fmt.Sprintf("🎧 %s's Music Wrapped", user.Username),
		Color:     style.ColorBrand,
		Footer:    style.Footer(),
		Thumbnail: &discordgo.MessageEmbedThumbnail{URL: user.AvatarURL("")},
		Fields: []*discordgo.MessageEmbedField{
			{Name: "Total Listening Time", Value: formatPlaytime(stats.TotalSeconds), Inline: true},
			{Name: "Tracks Played", Value: fmt.Sprintf("%d", stats.TotalTracks), Inline: true},
			{Name: "Unique Tracks", Value: fmt.Sprintf("%d", stats.UniqueTracks), Inline: true},
			{Name: "🎵 Top Tracks", Value: trackLines(topTracks)},
			{Name: "🎤 Top Artists", Value: artistLines(topArtists)},
		},
	}
	if hour >= 0 {
		embed.Fields = append(embed.Fields, &discordgo.MessageEmbedField{
			Name: "⏰ Peak Listening Hour", Value: formatHour(hour), Inline: true,
		})
	}
	return Respond(s, i, embed)
}

func wrappedServer(ctx context.Context, d Deps, s *discordgo.Session, i *discordgo.InteractionCreate) error {
	stats, err := d.Store.GuildWrappedStats(ctx, i.GuildID)
	if err != nil {
		return err
	}
	if stats.TotalTracks == 0 {
		return Respond(s, i, style.ErrorEmbed("No listening data found for this server yet!"))
	}

	topTracks, err := d.Store.TopTracks(ctx, i.GuildID, "", 5)
	if err != nil {
		return err
	}
	listeners, err := d.Store.MusicLeaderboard(ctx, i.GuildID, 5)
	if err != nil {
		return err
	}

	var listenerLines []string
	for idx, l := range listeners {
		listenerLines = append(listenerLines, fmt.Sprintf("%s <@%s> — %s (%d plays)",
			medal(idx), l.UserID, formatPlaytime(l.TotalSeconds), l.PlayCount))
	}
	listenersText := "No data"
	if len(listenerLines) > 0 {
		listenersText = strings.Join(listenerLines, "\n")
	}

	embed := &discordgo.MessageEmbed{
		Title:  fmt.Sprintf("🎧 %s Music Wrapped", guildName(s, i.GuildID)),
		Color:  style.ColorBrand,
		Footer: style.Footer(),
		Fields: []*discordgo.MessageEmbedField{
			{Name: "Total Listening Time", Value: formatPlaytime(stats.TotalSeconds), Inline: true},
			{Name: "Tracks Played", Value: fmt.Sprintf("%d", stats.TotalTracks), Inline: true},
			{Name: "Unique Artists", Value: fmt.Sprintf("%d", stats.UniqueArtists), Inline: true},
			{Name: "🎵 Top Tracks", Value: trackLines(topTracks)},
			{Name: "👑 Top Listeners", Value: listenersText},
		},
	}
	if g, err := s.State.Guild(i.GuildID); err == nil && g.Icon != "" {
		embed.Thumbnail = &discordgo.MessageEmbedThumbnail{URL: g.IconURL("")}
	}
	return Respond(s, i, embed)
}

func wrappedCompare(ctx context.Context, d Deps, s *discordgo.Session, i *discordgo.InteractionCreate, user, other *discordgo.User) error {
	if other.ID == user.ID {
		return Respond(s, i, style.ErrorEmbed("You can't compare with yourself!"))
	}

	type userData struct {
		stats   *storage.WrappedStats
		tracks  []storage.TrackPlays
		artists []storage.ArtistPlays
	}
	load := func(uid string) (*userData, error) {
		stats, err := d.Store.UserWrappedStats(ctx, i.GuildID, uid)
		if err != nil {
			return nil, err
		}
		tracks, err := d.Store.TopTracks(ctx, i.GuildID, uid, 50)
		if err != nil {
			return nil, err
		}
		artists, err := d.Store.TopArtists(ctx, i.GuildID, uid, 30)
		if err != nil {
			return nil, err
		}
		return &userData{stats: stats, tracks: tracks, artists: artists}, nil
	}

	u1, err := load(user.ID)
	if err != nil {
		return err
	}
	u2, err := load(other.ID)
	if err != nil {
		return err
	}
	if u1.stats.TotalTracks == 0 || u2.stats.TotalTracks == 0 {
		return Respond(s, i, style.ErrorEmbed("Both users need listening history to compare."))
	}

	names := func(artists []storage.ArtistPlays) []string {
		out := make([]string, len(artists))
		for idx, a := range artists {
			out[idx] = a.Artist
		}
		return out
	}
	percent, sharedArtists := compatibility(names(u1.artists), names(u2.artists))

	trackKeys := func(tracks []storage.TrackPlays) map[string]bool {
		m := make(map[string]bool, len(tracks))
		for _, t := range tracks {
			m[strings.ToLower(t.Title+"::"+t.Artist)] = true
		}
		return m
	}
	t1, t2 := trackKeys(u1.tracks), trackKeys(u2.tracks)
	sharedTracks := 0
	for k := range t1 {
		if t2[k] {
			sharedTracks++
		}
	}

	sharedText := "None"
	if len(sharedArtists) > 0 {
		show := sharedArtists
		if len(show) > 10 {
			show = show[:10]
		}
		sharedText = strings.Join(show, ", ")
	}

	embed := &discordgo.MessageEmbed{
		Title:       fmt.Sprintf("%s Music Compatibility: %d%%", compatEmoji(percent), percent),
		Description: fmt.Sprintf("**%s** vs **%s**", user.Username, other.Username),
		Color:       style.ColorBrand,
		Footer:      style.Footer(),
		Fields: []*discordgo.MessageEmbedField{
			{Name: user.Username, Value: fmt.Sprintf("%s\n%d plays", formatPlaytime(u1.stats.TotalSeconds), u1.stats.TotalTracks), Inline: true},
			{Name: other.Username, Value: fmt.Sprintf("%s\n%d plays", formatPlaytime(u2.stats.TotalSeconds), u2.stats.TotalTracks), Inline: true},
			{Name: "​", Value: "​", Inline: true},
			{Name: "🎤 Shared Artists", Value: sharedText},
			{Name: "📊 Stats", Value: fmt.Sprintf("**%d** shared artists • **%d** shared tracks", len(sharedArtists), sharedTracks)},
		},
	}
	return Respond(s, i, embed)
}
